### PR TITLE
Feature/tooltips for truncated labels

### LINKED ISSUE #15026: Implement hover hints for cramped text lines in UI
Due to the legacy driven interface limits a lot of titles and captions get cut by UI borders, and there is no possibility to see the ending of the title.
Would be nice to have a hint popup for such text showing the full line. Notable examples are server titles and replay file names. Missions window and assets browser are prone to have the same problem.

### LINKED ISSUE #16383: Quality of life improvements for support powers controls
This is actually a many-fold request, but I'll leave it like this for now just not to forget, if it gets worked on I can split it.

As per @WhoCares

- 1 key to cycle support powers (resetting cycle each time a power is applied or cancelled)
- support power have priority order in the cycle (IC and chrono always on top, in case of double take arbitrary decision of the one on top)
- While cycling, cameo appearing with the cursor giving you direct visual of what SP you're about to use/click

Practical scenario:
|| 3 support power case (chrono, dome, nuke)
press "g" key once : chronosphere
press "g" key twice : sonar pulse
press "g" key 3 times : nuke 
...
| 5 support power case 2 (airfield + IC)
press "g" key once : IC
press "g" key twice : parabomb
press "g" key 3 times : paradrop
press "g" key 4 times : spyplane
press "g" key 5 times : nuke

### PR DESCRIPTION
**Pull request by dragunoff**
Truncate text labels that are overflowing their containers and add tooltips to them that display the full text.

Affected panels:
- Savegame browser
- Assets browser
- Replay browser
- Mission browser
- Music player
- Lobby music player
- Server list
- Lobby server list

Notes:
- In a few places I had to add TooltipContainer with a unique name because the tooltips were rendering behind the panel
- ~~Also fixes a bug where tooltips in `lobby-servers.yaml` were not displayed at all~~ (no bug here, that was my bad)

A few screens:
replay-browser
Replay browser

lobby-server-list
Lobby server list

muisc-player
Music player

TESTCASE changes the width of some containers to cause an overflow and adds long text to music and mission panels. To test savegame browser type a long name for the save file.

Closes #15026

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/GameSaveBrowserLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/GameSaveBrowserLogic.cs]: This pattern could be pulled out into a `WidgetUtils.TruncatedLabelWithTooltip(LabelWidget label, string text)` to reduce duplication.

File: OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs]: Would you remind removing this long-redundant compiler workaround while you're here? Assign to `song` directly in the loop.

Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs]: Yes, I saw that in several other places (players?, maps?) as well and I was wondering why it's made like that. I'll change it.

Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs]: See https://blogs.msdn.microsoft.com/ericlippert/2009/11/12/closing-over-the-loop-variable-considered-harmful/ for context

File: OpenRA.Mods.Common/Widgets/Logic/GameSaveBrowserLogic.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/GameSaveBrowserLogic.cs]: I'm trying to do that but I'm getting an error:
```
The type or namespace name 'LabelWidget' could not be found (are you missing a using directive or an assembly reference?) (CS0246) [OpenRA.Game]
```

Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/GameSaveBrowserLogic.cs]: I see. This is because `WidgetUtils` is in the `OpenRA.Game` project, while all the widget code is in `OpenRA.Mods.Common`.  It looks like `GetContrastColor` is the only helper used in `OpenRA.Game`, and that is only used from `SpriteFont`.  If you move that into a private helper inside `SpriteFont` then `WidgetUtils` can move to `OpenRA.Mods.Common` and this will then work.

**Comment by dragunoff** [comment]
Updated to address the requested changes.

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: All this should go in the helper too!

**Comment by pchote** [comment]
Can you please reorder the commits so that the `WidgetUtils` changes are done first, and then the label/tooltip changes at the end use the new helper directly?

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: Got it. But then the helper had to return the whole widget. And to attach the new label in place of the old one I added a `ReplaceChild` method to `Widget`.

**Comment by dragunoff** [comment]
Updated with new changes and reordered commits.

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs
Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: I don't understand this. `LabelWidget` is a reference type, so there should be no need to return its value or replace it - it doesn't change!

Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: I'm sorry, learning C# as I go... I'll fix that later.

File: OpenRA.Mods.Common/Widgets/WidgetUtils.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/WidgetUtils.cs]: On second thought, `TruncateLabelToTooltip` might be a better name for this.

Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/WidgetUtils.cs]: Changed.

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: Fixed.

File: OpenRA.Mods.Common/Widgets/WidgetUtils.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/WidgetUtils.cs]: This needs to change to `OpenRA.Mods.Common.Widgets` as this is no longer in `OpenRA.Game`. The `using` above can then also go away, and a few other classes will need to add `using OpenRA.Mods.Common.Widgets`.

File: mods/cnc/chrome/gamesave-browser.yaml
Comment (pchote) [inline mods/cnc/chrome/gamesave-browser.yaml]: This should be able to stay as `TOOLTIP_CONTAINER` - the indentation change should be sufficient.

File: mods/cnc/chrome/music.yaml
Comment (pchote) [inline mods/cnc/chrome/music.yaml]: Please keep the end-of-file newline here, same comment on the ID naming

File: mods/common/chrome/assetbrowser.yaml
Comment (pchote) [inline mods/common/chrome/assetbrowser.yaml]: Please keep the end-of-file newline here.

File: mods/common/chrome/gamesave-browser.yaml
Comment (pchote) [inline mods/common/chrome/gamesave-browser.yaml]: Same comment on the ID naming.

File: mods/common/chrome/musicplayer.yaml
Comment (pchote) [inline mods/common/chrome/musicplayer.yaml]: end-of-file newline, same comment on ID naming.

File: mods/cnc/chrome/lobby-servers.yaml
Comment (pchote) [inline mods/cnc/chrome/lobby-servers.yaml]: Why change this? The original value should already have the correct margins wrt the password/auth icons.

File: mods/common/chrome/gamesave-browser.yaml
Comment (pchote) [inline mods/common/chrome/gamesave-browser.yaml]: Same comment on ID naming.

File: mods/common/chrome/musicplayer.yaml
Comment (pchote) [inline mods/common/chrome/musicplayer.yaml]: Same comment on ID naming.

File: mods/cnc/chrome/lobby-servers.yaml
Comment (dragunoff) [reply mods/cnc/chrome/lobby-servers.yaml]: This is part of the TESTCASE

File: mods/cnc/chrome/gamesave-browser.yaml
Comment (dragunoff) [reply mods/cnc/chrome/gamesave-browser.yaml]: It's fine in the menu but it renders below the panel when in-game.

File: mods/common/chrome/assetbrowser.yaml
Comment (dragunoff) [reply mods/common/chrome/assetbrowser.yaml]: EOF lines fixed everywhere.

File: OpenRA.Mods.Common/Widgets/WidgetUtils.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/WidgetUtils.cs]: Fixed.

**Review by pchote — DISMISSED** [review]
LGTM and works as advertised. Testcase will need to be removed before merging.

**Review by abcdefg30 — DISMISSED** [review]

**Comment by abcdefg30** [comment]
(This can be merged once the testcase is gone.)

**Comment by dragunoff** [comment]
Testcase removed.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/70a2a0ba82c71dd95410bff5017f1cfc66032a5e...e22de4296b9c66c327ee47f1b2c553518cbcbd34)

**Pull request merged**
